Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,213 @@
+
+variable {R S X}
+
+/-- uncurry a map from a tensor product to a bilinear map-/
+noncomputable def curry' {X' : ModuleCat.{v} R} {Y : ModuleCat.{v} Sᵐᵒᵖ}
+    (l : (tensorFunctor R S X).obj X' ⟶ Y) :
+    X' ⟶ (homFunctor R S X).obj Y where
+  toFun x' :=
+    { toFun := (l <| . ⊗ₜ x')
+      map_add' := fun _ _ => show l _ = l _ + l _ by rw [add_tmul _ _ x', map_add]
+      map_smul' := fun s _ => show l _ = s • l _ by rw [← smul_tmul', map_smul] }
+  map_add' := fun _ _ => LinearMap.ext fun _ => show l _ = l _ + l _ by rw [tmul_add, map_add]
+  map_smul' := fun r _ => LinearMap.ext fun _ => show l _ = l (r • _ ⊗ₜ _) by
+    rw [← smul_tmul, smul_tmul']
+
+attribute [aesop unsafe] add_comm in
+/-- curry a bilinear map into a map from tensor product -/
+@[simps]
+def uncurry' {X' : ModuleCat.{v} R} {Y : ModuleCat.{v} Sᵐᵒᵖ} (l : X' →ₗ[R] (X →ₗ[Sᵐᵒᵖ] Y)) :
+    (X ⊗[R] X') →ₗ[Sᵐᵒᵖ] Y :=
+  let L : (X ⊗[R] X') →+ Y := (addConGen _).lift (FreeAddMonoid.lift fun p => l p.2 p.1) <|
+    AddCon.addConGen_le <| by
+      rintro _ _ (_|_|_|_|_|_) <;> refine' (AddCon.ker_rel _).2 _ <;> try aesop
+      simpa only [FreeAddMonoid.lift_eval_of, l.map_smul] using
+        Eq.symm <| LinearMap.bimodule_smul_apply (MulOpposite.op _) (l _) _
+  { L with
+    map_smul' := fun _ z => z.induction_on
+      (by aesop) (fun _ _ => LinearMap.map_smul _ _ _) (by aesop) }
+
+@[simp high]
+lemma uncurry'_apply_tmul {X' : ModuleCat.{v} R} {Y : ModuleCat.{v} Sᵐᵒᵖ}
+    (l : X' →ₗ[R] (X →ₗ[Sᵐᵒᵖ] Y))
+    (x : X) (x' : X') : uncurry' l (x ⊗ₜ x') = l x' x := rfl
+
+variable (R S X)
+/-- The tensor functor is left adjoint to the hom functor. -/
+noncomputable def tensorHomAdjunction : tensorFunctor R S X ⊣ homFunctor R S X :=
+    Adjunction.mkOfHomEquiv
+    { homEquiv := fun _ _ =>
+      { toFun := curry'
+        invFun := uncurry'
+        left_inv := fun _ => LinearMap.ext fun x => x.induction_on (by aesop) (by aesop) (by aesop)
+        right_inv := fun _ => LinearMap.ext fun _ => LinearMap.ext fun _ => rfl }
+      homEquiv_naturality_left_symm := fun {_ _} _ _ _ => by
+        refine' LinearMap.ext fun x => x.induction_on _ _ _ <;> aesop
+      homEquiv_naturality_right := by aesop }
+
+noncomputable instance : IsLeftAdjoint (tensorFunctor R S X) :=
+⟨_, tensorHomAdjunction _ _ _⟩
+
+noncomputable instance : IsRightAdjoint (homFunctor R S X) :=
+⟨_, tensorHomAdjunction _ _ _⟩
+
+noncomputable instance : Limits.PreservesColimits (tensorFunctor R S X) :=
+Adjunction.leftAdjointPreservesColimits <| tensorHomAdjunction _ _ _
+
+noncomputable instance : Limits.PreservesLimits (homFunctor R S X) :=
+Adjunction.rightAdjointPreservesLimits <| tensorHomAdjunction _ _ _
+
+instance : Functor.PreservesEpimorphisms (tensorFunctor R S X) :=
+inferInstance
+
+end ModuleCat
+
+namespace TensorProduct
+
+variable (R' : Type u) [CommRing R']
+variable {M N : Type v} [AddCommGroup M] [AddCommGroup N]
```

Comment:
These should not be constrained to the same universe

Reply:
In commite [63ec5b0](https://github.com/leanprover-community/mathlib4/pull/8495/commits/63ec5b0710f344c77c1bf2bb16fb0c60f39cfe85) I have generalized the universe levels